### DISCUSSION TITLE
weird code in the immediatelyRunFunctionInMainWorld in the file content-scripts/prototype-handler.js

### DISCUSSION THREAD

**Discussion by OriLiMu** [ScratchAddons/ScratchAddons#7543] [Help] [answered]
The code is like 
```js
function immediatelyRunFunctionInMainWorld(fn) {
  if (typeof fn !== "function") throw "Expected function";
  const div = document.createElement("div");
  div.setAttribute("onclick", "(" + fn + ")()");
  document.documentElement.appendChild(div);
  div.click();
  div.remove();
}
```
Really weird that, create an div then click it then remove it. 
Why not just call the function bind to `onclick`? 
What is the meaning of the add/remove element work? 
And besides I remember that the guidance of the dev doc not recommend to use the `onclick` thing, use `addListener` is the best. 
So Anyone can explain this to me?

**Comment by CST1229** [comment]
i think this is a workaround to run main world scripts in content scripts, because they run in a separate context and the csp they have don't let you add arbitrary `<script>` tags (but setting an `onclick` then calling `.click()` works)

**Reply by OriLiMu** [reply]
Hi I am new to the scratchAddon. And I am reading the code now. 
Can you explain some terminology for me. 
1. what is content-scripts used for, I see two `content-scripts` folder one in the root, one in the addons-api folder? what is the difference? 
2. what is vm, I see a lot of `vm.runtime` in the code. vm is virtual environment?
3. what is MAIN_WORLD. I read this from code comment
> // Unfortunately `chrome.scripting.registerContentScripts` can only be set to **MAIN_WORLD**
> // since Chrome 102, and we want to support Chrome 96+. 
4. What is trap? the explain from the doc is blur: 
> There are two types of traps, named after how many times they get trapped. Singleton instance like VM or ScratchBlocks get trapped only once per trap, so they are called “Once”. Objects like Thread can get trapped multiple times, so they are called “Many”.

I still don't understand the concept of trap after reading this

**Reply by mxmou** [reply, marked as answer]
> 1. what is content-scripts used for, I see two `content-scripts` folder one in the root, one in the addons-api folder? what is the difference?

Content scripts are JavaScript files which a browser extension can use to interact with web pages: https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts

Scratch Addons' content scripts are inside the `content-scripts` directory in the root. `addon-api/content-script` is unrelated - it contains those parts of the [addon API](https://scratchaddons.com/docs/reference/addon-api/) that are only available to addons running on the Scratch page.

> 2. what is vm, I see a lot of `vm.runtime` in the code. vm is virtual environment?

`vm` is an instance of [`VirtualMachine`](https://github.com/scratchfoundation/scratch-vm/blob/develop/src/virtual-machine.js). It's essentially a representation of a Scratch project.

> 3. what is MAIN_WORLD. I read this from code comment
> 
> > // Unfortunately `chrome.scripting.registerContentScripts` can only be set to **MAIN_WORLD**
> > // since Chrome 102, and we want to support Chrome 96+.

By default, content scripts run in a separate "world" from the Scratch page, which means that global variables in one world aren't accessible in the other. Running in the "main world" allows addons to interact with Scratch's code.

> 4. What is trap? the explain from the doc is blur:
> 
> > There are two types of traps, named after how many times they get trapped. Singleton instance like VM or ScratchBlocks get trapped only once per trap, so they are called “Once”. Objects like Thread can get trapped multiple times, so they are called “Many”.
> 
> I still don't understand the concept of trap after reading this

I think this page on the docs is outdated - there are no "many" traps in the current version of the extension.

The `traps` API allows addons to access objects defined in Scratch's code. The name "traps" is from earlier versions which used to [override some built-in JS methods](https://github.com/apple502j/ScratchAddons/blob/701c60908927d61ab2763cdd8deddaad7cbfdf28/content-scripts/prototype-handler.js) to "trap" such objects.

**Comment by WorldLanguages** [comment]
> And besides I remember that the guidance of the dev doc not recommend to use the `onclick` thing, use `addListener` is the best.

That guidance is for addon code, not for all code in the extension.
Most contributors never read a single line of code outside of the /addons/ directory. It's not really necessary to contribute new features, addons, and bug fixes to addons.
